File: src/MatchingAlgorithm/HeatingSystem.cpp
#include "HeatingSystem.h"

#include <complex>

HeatingSystem::HeatingSystem(const std::vector<HeatingSystemData>& frequency, const std::vector<HeatingSystemData>&
							 temperature)
{
	for(auto [Key, Resistance, Inductance] : frequency)
	{
		_resistanceFrequency.insert({Key, Resistance});
		_inductanceFrequency.insert({Key, Inductance});
	}

	for(auto [Key, Resistance, Inductance] : temperature)
	{
		_resistanceTemperature.insert({Key, Resistance});
		_inductanceTemperature.insert({Key, Inductance});
	}
}


double HeatingSystem::Resistance(const double frequency, const double temperature) const
{
	return _resistanceFrequency.at(frequency) * _resistanceTemperature.at(temperature);
}

double HeatingSystem::Inductance(const double frequency, const double temperature) const
{
	return _inductanceFrequency.at(frequency) * _inductanceTemperature.at(temperature);
}

double HeatingSystem::Impedance(const double frequency, const double temperature) const
{
	return abs(std::complex{Resistance(frequency, temperature), Inductance(frequency, temperature)});
}

File: src/MatchingAlgorithm/HeatingSystem.cs
using System.Numerics;

namespace MatchingAlgorithm;

public class HeatingSystem : IHeatingSystem
{
    private readonly Dictionary<double, double> _resistanceFrequency;
    private readonly Dictionary<double, double> _inductanceFrequency;
    private readonly Dictionary<double, double> _resistanceTemperature;
    private readonly Dictionary<double, double> _inductanceTemperature;

    public HeatingSystem(IEnumerable<HeatingSystemData> frequency, IEnumerable<HeatingSystemData> temperature)
    {
        var fList = frequency.ToArray();
        var tList = temperature.ToArray();
        _resistanceFrequency = fList.ToDictionary(key => key.Key, value => value.Resistance);
        _inductanceFrequency = fList.ToDictionary(key => key.Key, value => value.Inductance);
        _resistanceTemperature = tList.ToDictionary(key => key.Key, value => value.Resistance);
        _inductanceTemperature = tList.ToDictionary(key => key.Key, value => value.Inductance);
    }

    public double Resistance(double frequency, double temperature)
    {
        return _resistanceFrequency[frequency] * _resistanceTemperature[temperature];
    }

    public double Reactance(double frequency, double temperature)
    {
        return _inductanceFrequency[frequency] * _inductanceTemperature[temperature];
    }

    public Complex Impedance(double frequency, double temperature)
    {
        return new Complex(Resistance(frequency, temperature), Reactance(frequency, temperature));
    }
}